src/mlir/dialect/VeronaOps.cc

```diff
@@ -4,7 +4,114 @@
 #include "VeronaOps.h"
 
 #include "VeronaDialect.h"
+#include "mlir/IR/Builders.h"
 #include "mlir/IR/OpImplementation.h"
+#include "mlir/IR/StandardTypes.h"
+
+#include "llvm/ADT/StringSet.h"
+
+using namespace mlir;
+
+static ParseResult parseClassOp(OpAsmParser& parser, OperationState& state)
+{
+  Region* body = state.addRegion();
+
+  StringAttr nameAttr;
+  if (parser.parseSymbolName(
+        nameAttr, SymbolTable::getSymbolAttrName(), state.attributes))
+    return failure();
+
+  if (parser.parseOptionalAttrDictWithKeyword(state.attributes))
+    return failure();
+
+  if (parser.parseRegion(*body, /*arguments=*/{}, /*argTypes=*/{}))
+    return failure();
+
+  verona::ClassOp::ensureTerminator(*body, parser.getBuilder(), state.location);
+
+  return success();
+}
+
+static void print(OpAsmPrinter& printer, verona::ClassOp op)
+{
+  printer << verona::ClassOp::getOperationName() << ' ';
+  printer.printSymbolName(op.sym_name());
+  printer.printOptionalAttrDict(
+    op.getAttrs(), /*elidedAttrs=*/{SymbolTable::getSymbolAttrName()});
+  printer.printRegion(
+    op.body(), /*printEntryBlockArgs=*/false, /*printBlockTerminators=*/false);
+}
+
+/**
+ * AllocateRegionOp and AllocateObjectOp share similar features, which we verify
+ * here:
+ * - The class referenced by the operation must exist.
+ * - There must be as many field names as there are operands.
+ * - The set of field names must match those specified in the class definition (TODO: not implemented yet)
+ *
+ * TODO: Rather than a templated function, should this be a trait?
```

Comment:
This looks like a pretty standard verify pass, to me. A trait is more about the operation itself, not the state of the arguments outside the operation.